### PR TITLE
 Composite primary key & association entities as keys improvements

### PR DESCRIPTION
**Pull request by goetas**
Hi! 
I tried to implement the matching of entities that uses composite primary keys...

Any suggestion?

~~BC changes here https://github.com/doctrine/doctrine2/pull/1113/files#diff-c18001a2ca9928743b1e99854ffad33bL1611 (protected method makred as private and changed return type..)~~

### CODE REVIEW

**Comment by goetas** [comment]
@Ocramius any chance to see this merged before 3.0?

**Comment by Ocramius** [comment]
@goetas this will land in 2.5

**Comment by Ocramius** [comment]
I can't deal with the rebase myself tonight, sorry :-\

**Comment by goetas** [comment]
I will tomorrow

**Comment by goetas** [comment]
This PR https://github.com/doctrine/doctrine2/pull/1178 changed (breaking) almost everything... :-(

Now it is not easy to fix it.... Since this is not a new feature, i hope that it can be ingratiated in some future minor release...

The `PersisterHelper` in my opinion it is a very bad idea... static helper are usually dangerous and tends to be completely procedural... as example `getIndividualValue` is already (partially) handled by `IdentifierFlattener` class...

**Comment by goetas** [comment]
@Ocramius can you just pick this https://github.com/goetas/doctrine2/commit/b01d1810e5164d166ee28bdae6a23e9010591bd8 commit and mark the mentioned method as private before releasing 2.5?

Since I will continue to work on this in the next week...

**Comment by Ocramius** [comment]
@goetas will on tuesday, I think.

Yes, #1178 caused a big rebase mess :-\

**Comment by goetas** [comment]
It does not pass the tests suite, but this is one step forward!

**Comment by Ocramius** [comment]
Awesome, and thanks for keeping hammering on it!

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 1]
```diff
@@ -864,12 +868,12 @@ public function expandCriteriaParameters(Criteria $criteria)
...
         foreach ($params as $param) {
-            $sqlParams[] = PersisterHelper::getIdentifierValues($param, $this->em);
+            $sqlParams = array_merge($sqlParams, $this->getValues($param));
         }
 
         foreach ($types as $type) {
-            list($field, $value) = $type;
-            $sqlTypes[]          = $this->getType($field, $value, $this->class);
+            list ($field, $value) = $type;
+            $sqlTypes = array_merge($sqlTypes, $this->getTypes($field, $value, $this->class));
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]: Types seem to be incorrectly bound here, as the value conversion tests (for association columns) fail.

**Comment by goetas** [comment]
@Ocramius i'm still working on this. I do not know if is a good "temporary solution" but i have created #1292 to mark as private the `getSelectConditionStatementColumnSQL`. This will avoid BC break in 2.5

**Comment by goetas** [comment]
Now test suite should pass! :-D

I have to rebase it and fix some  issues.

I also think that i have found some kind of bug with second level cache.
[Here](https://github.com/doctrine/doctrine2/blob/17a23ea82595d49995f347bf6dae7e218e98f1f4/lib/Doctrine/ORM/Cache/DefaultCollectionHydrator.php#L92)` $entityEntry->data` can contain objects (used by SLC to track associations), but UnitOfWork [here](https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/UnitOfWork.php#L2498) supposes to have always scalars.
BTW I'm planning to fix also this.

**Comment by Ocramius** [comment]
Awesome! I'll review it tomorrow :-)

**Comment by goetas** [comment]
@Ocramius The PR is still a work in progress, most probably this week I will fix some known problems, and later you can review it. Can it work for you?

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 2]
```diff
@@ -2639,10 +2625,18 @@ public function createEntity($className, array $data, &$hints = array())
 
                         if ($joinColumnValue !== null) {
                             if ($targetClass->containsForeignIdentifier) {
+                                if ($joinColumnValue instanceof AssociationCacheEntry) {
```
Comment (goetas) [inline lib/Doctrine/ORM/UnitOfWork.php]: I'm really not proud of this. Ideas?

File: lib/Doctrine/ORM/Utility/IdentifierFlattener.php [hunk 3]
```diff
@@ -71,20 +71,29 @@ public function flattenIdentifier(ClassMetadata $class, array $id)
     {
         $flatId = array();
 
-        foreach ($id as $idField => $idValue) {
-            if (isset($class->associationMappings[$idField]) && is_object($idValue)) {
-                /* @var $targetClassMetadata ClassMetadata */
+        foreach ($class->identifier as $field) {
+            if (isset($class->associationMappings[$field]) && isset($id[$field]) && $id[$field] instanceof \Doctrine\ORM\Cache\AssociationCacheEntry) {
```
Comment (goetas) [inline lib/Doctrine/ORM/Utility/IdentifierFlattener.php]: Also not proud of this

**Comment by goetas** [comment]
Ok, now looks berrer

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 4]
```diff
@@ -335,7 +335,11 @@ protected function fetchVersionValue($versionedClass, $id)
              . ' FROM '  . $tableName
              . ' WHERE ' . implode(' = ? AND ', $identifier) . ' = ?';
 
-        $flatId = $this->identifierFlattener->flattenIdentifier($versionedClass, (array) $id);
+        if (!is_array($id)) {
+            $id = array($this->class->identifier[0] => $id);
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]: @goetas this is very naive, and assumes a single column identifier.

What if I try to fetch the version of an entity with a composite identity?

Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]: We may also avoid calling this method with a `mixed` `$id`

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 5]
```diff
@@ -1607,7 +1632,7 @@ public function getSelectConditionStatementSQL($field, $value, $assoc = null, $c
      * @param string     $field
      * @param array|null $assoc
      *
-     * @return string
+     * @return array
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]: `string[]`?

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 2 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: @FabioBatSilva maybe got ideas?

File: tests/Doctrine/Tests/ORM/Utility/IdentifierFlattenerTest.php [hunk 6]
```diff
@@ -87,7 +87,7 @@ public function testFlattenIdentifierWithOneToOneId()
 
         $this->assertArrayHasKey('secondEntity', $flatIds, 'It should be called secondEntity');
 
-        $this->assertSame($id['secondEntity']->id, $flatIds['secondEntity']);
+        $this->assertEquals($id['secondEntity']->id, $flatIds['secondEntity']);
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Utility/IdentifierFlattenerTest.php]: @goetas makes sense, though we really should stop doing that unless it's a hashed ID. I'll think about it, thanks.

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 2 above)
Comment (goetas) [inline lib/Doctrine/ORM/UnitOfWork.php]: @Ocramius It was my mistake... fixed with https://github.com/goetas/doctrine2/commit/268426aa85e56a7bf487dd706e81ba24764b8393
/cc @FabioBatSilva

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php (hunk 4 above)
Comment (goetas) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]: Changing the `assignDefaultVersionValue` signature, simplified a lot! Good advice! :-D

**Comment by goetas** [comment]
Now I'm more satisfied with the pull request status!

**Comment by Ocramius** [comment]
Manually merged after a rebase at doctrine/doctrine2@f90897465cb18d1aed0150dd67d962b956bd1c2a

**Comment by Ocramius** [comment]
Thanks @goetas, as usual nice work :-)

**Pull request closed without merging**